### PR TITLE
feat: implement global decorator config

### PR DESCRIPTION
**Pull request by taehee-sp**
issue https://github.com/cjohansen/portfolio/issues/28
 
I implemented global decorator support for portfolio.

here is current config api (can be changed)

```clojure
(defonce app
  (ui/start!
	@@ -31,6 +32,9 @@

    :config
    {:css-paths ["/portfolio/demo.css"]
     :decorators {:reagent    reagent-decorator
                  :reagent-18 reagent-decorator
                  :react-18   react-18-decorator}
   }))
```

example decorator implementation.

```clojure
(defn reagent-decorator [child]
  (createElement Provider #js{:value :red}
                 (r/as-element child)))

(defn react-18-decorator [props]
  (createElement Provider #js{:value :red}
                 (.-children props)))
```

here is consumer example scene.

```clojure
(defn decorator-consumer-component []
  (let [theme (use-theme)]
    ($ :button {:style {:background (name theme)}}
       "current theme is " (name theme))))

(defscene decorator-consumer-demo
  :title "using global decorator context"
  ($ decorator-consumer-component))
```

## Issue
I just read decorators config from `@app` atom. I know it's not ideal, but it's difficult for me to understand the whole structure. Can you give me a suggestion please?

## TODO
- [ ] add decorator docs
- [ ] react-toastify, @tanstack/query decorator example.

### CODE REVIEW

**Comment by cjohansen** [comment]
I agree that accessing the `app` atom this way isn't ideal. I don't think it's necessary to add the decorators to the app state at all, since they are not data. The decorators are more a detail of the rendering mechanics, like the individual framework-specific frameworks. So my suggestion is to go with a more code-driven solution, something like:

```clj
(require '[portfolio.reagent :as pr])

(pr/set-decorator! reagent-decorator)
```

This will just be a convention followed in each adapter, and can be implemented like so:

```clj
(ns portfolio.reagent
  ,,,)

(def *decorator* nil)

(defn set-decorator! [decorator]
  (set! *decorator* decorator))
```

Then:

```clj
;; Replace this
(let [decorator (or (:reagent (:decorators @app)) identity)] ,,,)

;; ...with
(let [decorator (or *decorator* identity)] ,,,)
```

A note on code-style: I see a lot of the namespace forms have been reformatted in various ways. This makes it hard to gauge what changes you've made. Some of them are reasonable enough, but I'd prefer the PR to only have the minimal changes necessary to support your feature, or at the very least separate commits for stuff like formatting. I also prefer using `:as` when requiring namespaces, and not using`:refer`, as I find it makes the code much harder follow. I've used `:refer-macros` sparingly for things like `defscene` etc, but I don't want Portfolio's internals to use `:refer` for regular functions.

**Comment by taehee-sp** [comment]
Thank you for the comments. I understand the points.

I will change the PR as you said. (just `pr/set-decorator!` and use `:as` instead of `:refer`)

**Comment by taehee-sp** [comment]
Just a friendly reminder. @cjohansen

File: sample/src/portfolio/components/helix.cljs [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
+            [helix.dom :as d] ;; If you are using an older version of react use the following:
             [helix.hooks :as hooks]
-            [helix.dom :as d]
-            ;; If you are using an older version of react use the following:
-            #_[portfolio.react :refer-macros [defscene]]
-            ;; For react versions 18+ use the following:
-            [portfolio.react-18 :refer-macros [defscene]]
-            ["react" :as react]))
+            [portfolio.theme :as theme]
+            [portfolio.react-18 :refer-macros [defscene]]))
```
Comment (cjohansen) [inline sample/src/portfolio/components/helix.cljs:10]: Please undo these unrelated changes.

File: sample/src/portfolio/components/helix.cljs [hunk 2]
```diff
@@ -30,11 +29,11 @@
...
-      (render [this]
-        (when (.. this -props -error)
-          (throw (js/Error. "BOOOOOM!")))
-        "Oh, nice!!"))
+              Object
+              (render [this]
+                      (when (.. this -props -error)
+                        (throw (js/Error. "BOOOOOM!")))
+                      "Oh, nice!!"))
     ctor))
```
Comment (cjohansen) [inline sample/src/portfolio/components/helix.cljs:37]: This code seems to have been reformatted with no real changes. Please undo it.

File: sample/src/portfolio/components/reagent.cljs [hunk 3]
```diff
@@ -4,7 +4,8 @@
   ;; For react versions 18+ use the following:
   ;; This is due to the new API https://www.metosin.fi/blog/reagent-towards-react-18/
   (:require [portfolio.components.reagent.component :as rc]
-            [portfolio.reagent-18 :refer-macros [defscene]]))
+            [portfolio.theme :as theme]
+            [portfolio.reagent :refer-macros [defscene]]))
```
Comment (cjohansen) [inline sample/src/portfolio/components/reagent.cljs:9]: Why did you change the React version in use here?

**Review by cjohansen — CHANGES_REQUESTED** [review]
Sorry for the very delayed response!

I like this API a lot better. It still needs to be defined for the remaining non-React adapters, but I can do that. They will probably just be function wrappers.

As I mentioned in the initial comment, there is a lot of unrelated and unnecessary changes to the code, mainly formatting and ordering. I would appreciate if you could revert these, I've highlighted the individual places where I want to keep the original formatting.

File: sample/src/portfolio/components/reagent.cljs (hunk 3 above)
Comment (taehee-sp) [reply sample/src/portfolio/components/reagent.cljs:9]: it's maybe lsp related mistake. i will revert it.

File: sample/src/portfolio/components/rum.cljs [hunk 4]
```diff
@@ -1,14 +1,25 @@
...
-            [portfolio.react-18 :refer-macros [defscene]]))
+  (:require #_[portfolio.rum :refer-macros [defscene]] ;; For react versions 18+ use the following:
+ ;; Beware newer versions of react may not work as expected rum is
+ ;; not actively maintained https://github.com/tonsky/rum/issues/165
+ ;; For react versions 18+ use the following:
+ ;; Beware newer versions of react may not work as expected rum is
+   [portfolio.theme :as theme]
+            [portfolio.react-18 :refer-macros [defscene]]
+            [rum.core :as rum] ;; If you are using an older version of react use the following:
+            ))
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply sample/src/portfolio/components/rum.cljs:11]: i will undo it

File: sample/src/portfolio/components/uix.cljs [hunk 5]
```diff
@@ -1,14 +1,25 @@
 (ns portfolio.components.uix
-  (:require [uix.core :refer [defui $ use-state]]
-            [uix.dom]
-            [portfolio.react-18 :refer-macros [defscene]]))
+  (:require [portfolio.theme :as theme]
+            [portfolio.react-18 :refer-macros [defscene]]
+            [uix.core :refer [$ defui use-state]]
+            [uix.dom]))
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply sample/src/portfolio/components/uix.cljs:6]: I will undo it

File: sample/src/portfolio/scenes.cljs [hunk 6]
```diff
@@ -11,13 +11,21 @@
...
             [portfolio.layouts.tall-page]
             [portfolio.locale-selector :as locale-selector]
+            [portfolio.react-18 :as p-react-18]
+            [portfolio.reagent :as p-reagent]
+            [portfolio.reagent-18 :as p-reagent-18]
             [portfolio.ui :as ui]
-            [portfolio.ui.search :as search]
-            [portfolio.ui.search.memsearch-index :as memsearch]))
+            [portfolio.ui.search :as search]))
+
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply sample/src/portfolio/scenes.cljs:24]: I will undo it

File: sample/src/portfolio/scenes.cljs [hunk 7]
```diff
@@ -11,13 +11,21 @@
...
+(p-react-18/set-decorator! theme/react-18-decorator)
+
+(p-reagent/set-decorator! theme/reagent-decorator)
+(p-reagent-18/set-decorator! theme/reagent-decorator)
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply sample/src/portfolio/scenes.cljs:29]: okay! i will do it

File: sample/src/portfolio/scenes.cljs [hunk 8]
```diff
@@ -75,12 +83,12 @@
...
 
+(comment
   (->> [:set-css-files ["/portfolio/demo2.css"]]
        (portfolio.actions/execute-action! app))
 
   (->> [:set-css-files ["/portfolio/demo.css"]]
-       (portfolio.actions/execute-action! app))
-
-)
+       (portfolio.actions/execute-action! app)))
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply sample/src/portfolio/scenes.cljs:94]: i will undo it

File: src/portfolio/react.cljs [hunk 9]
```diff
@@ -8,15 +8,14 @@
...
-
 (def component-impl
   {`adapter/render-component
    (fn [scene el]
      (assert (some? el) "Asked to render component into null container.")
      (when-let [f (some-> el .-unmount)]
        (f))
-     (react-dom/render (react/createElement Wrapper #js {:scene scene}) el))})
+     (let [Wrapper (react-util/create-safe-wrapper)]
+       (react-dom/render (react/createElement Wrapper #js {:scene scene}) el)))})
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply src/portfolio/react.cljs:19]: oh it's because the Wrapper use *decorator* (which is mutable) so we have to recreate it at runtime.

File: src/portfolio/react_utils.cljs [hunk 10]
```diff
@@ -1,9 +1,14 @@
...
-  (:require [goog]
+  (:require ["react" :as react]
+            [goog]
             [goog.object :as o]
-            ["react" :as react]
-            [portfolio.ui.actions :as actions]
-            [portfolio.adapter :as adapter]))
+            [portfolio.adapter :as adapter]
+            [portfolio.ui.actions :as actions]))
+
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply src/portfolio/react_utils.cljs:7]: okay!

File: src/portfolio/reagent.cljs [hunk 11]
```diff
@@ -7,23 +7,31 @@
...
-               (fn []
-                 (rd/unmount-component-at-node el)
-                 (set! (.-unmount el) nil))))
-       (js/console.error "Asked to render Reagent component without an element")))})
+     (let [decorator (or *decorator*
+                         identity)]
+       (if el
+         (do
+           (rd/render [decorator
+                       (if (fn? component)
```
[the remark this answers is not in the document]
Comment (taehee-sp) [reply src/portfolio/reagent.cljs:26]: okay!

**Comment by taehee-sp** [comment]
it's done! please review @cjohansen  :)

**Comment by taehee-sp** [comment]
all done! (I checked my PR diff one by one) @cjohansen please review
